Extend Governor with parameterized votes

apbendi opened the pull request:
This PR lays the groundwork for a future implementation of #2929, along with other potentially useful Governance extensions, by making the `_castVote` interface more flexible. In particular, this PR follows @frangio's [suggestion](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2929#issuecomment-961999046) and adds a `data` parameter to the internal, virtual `_castVote` method signature, along with associated `countVote` methods.

The idea is this `data` field can be used to encode arbitrary data about the vote in question, enabling many interesting voting mechanisms via Governor extension. It can be coupled with a reporting of which kind of standardized voting mechanisms are supported in the `COUNTING_MODE` function.

I'm opening this PR as draft for two reasons:

1. To get feedback on the general approach, as suggested by @frangio and implemented here. It seems reasonable to me but I'm very open to feedback!
2. To get suggestions on better naming. `data` as a parameter name and `WithData` for method extensions seems very generic. I think we can do better but I don't have any great ideas. One possibility would be `countingContext`, `countingData`, or something like this. Any ideas?

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog entry

moodysalem commented on contracts/governance/compatibility/GovernorCompatibilityBravo.sol:
```diff
@@ -265,8 +265,11 @@ abstract contract GovernorCompatibilityBravo is IGovernorTimelock, IGovernorComp
+        uint256 weight,
+        bytes memory data
     ) internal virtual override {
+        (data); // silence compiler warnings
```
remove the parameter name?

apbendi replied:
I like having the param name for self documentation purposes, but not opposed to removing if this style is preferred.

apbendi commented on contracts/governance/Governor.sol:
```diff
@@ -67,7 +67,11 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
-        return interfaceId == type(IGovernor).interfaceId || super.supportsInterface(interfaceId);
+        // In addition to the current interfaceId, also support previous version of the interfaceId that did not
+        // include the castVoteWithData() and castVoteWithReasonAndData() functions as standard
+        return interfaceId == (type(IGovernor).interfaceId ^ this.castVoteWithReasonAndData.selector ^ this.castVoteWithData.selector)
```
One additional area where I would appreciate some feedback is on this approach. An alternative, to avoid breaking backwards compatibility, would be to define an additional interface, like `IGovernorWithData` (ugh, naming) or `IGovernorCountingContext` that extends from `IGovernor`, then change this to something like:

```solidity
return interfaceId == type(IGovernorCountingContext).interfaceId
    || interfaceId == type(IGovernor).interfaceId
    || super.supportsInterface(interfaceId);
```

frangio replied:
I think I'm fine with the way this is implemented currently.

frangio commented:
Thank you @apbendi! I agree that "data" is a bad name, I am thinking what we're adding here are "parameters" for the vote. So I'd go with "params", or if we want to tie it into the "counting" concept it could be "counting parameters". What do you think?

Have you tried to implement something like https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2929 on top of these changes? It would be good to see if it actually works well for that purpose.

What I think is missing is a virtual function `_defaultParams()` with a default return value of empty string, and we can use in place of the hardcoded empty string there is in the PR currently. A module like fractional voting would override `_defaultParams()` to return the value for "100%", is what I"m thinking.

apbendi commented:
Thanks for the quick reply @frangio!

>  I am thinking what we're adding here are "parameters" for the vote. So I'd go with "params"

This seems reasonable, and certainly an improvement over `data`. I'll make this change.

> Have you tried to implement something like #2929 on top of these changes? It would be good to see if it actually works well for that purpose.

I'm actively working on this, yes. Can give you some impressions as I get a little further.

> What I think is missing is a virtual function _defaultParams() with a default return value of empty string

Makes sense. I will add this.

Also, any opinion on [this](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043#discussion_r773460876).

Thanks!

frangio commented:
>Can give you some impressions as I get a little further.

Please do! We want to make sure it serves the intended purpose.

One more thing I wanted to suggest is just adding a single new function with both reason and params, so as not to explode the number of functions. We've also found that memory parameters in external/public functions contribute to bytecode size quite a bit.

apbendi commented:
hey @frangio, I believe I've addressed all feedback. I'm thinking I can take this PR out of Draft if that works for you. Would you like me to squash everything back to one commit? Also can you help me understand the codecov CI failure I'm seeing— looks like maybe the code coverage of the diff is not high enough, is that right?

frangio commented on contracts/governance/Governor.sol:
```diff
@@ -183,9 +188,20 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
         uint8 support,
-        uint256 weight
+        uint256 weight,
+        bytes memory params
```
Adding this argument is a breaking change. We'd like some way to make this backwards compatible. Think of custom counting modules that should continue to compile and work.

If we can't find a good way to make this work we can consider making the breaking change but we strongly prefer not to.

One way could be to keep the previous function signature, and add the new one with a default implementation that redirects to the old one. But then if one wants to implement a counting module using the new signature, what should they override? The new one, old one, or both, and then how? That doesn't make a lot of sense.

An alternative is to have some other contract `GovernorWithParams` where `Governor` is kept as the backwards compatible version. What I don't like is this adds quite some complexity and due to Solidity `override` syntax leaking implementation details it is probably a breaking change in another way.

apbendi replied:
Hey @frangio, hope you had a good holiday! Coming back to this now and reviewing what you said.

> One way could be to keep the previous function signature, and add the new one with a default implementation that redirects to the old one. But then if one wants to implement a counting module using the new signature, what should they override? The new one, old one, or both, and then how? That doesn't make a lot of sense.

Yeah this seems like it'd be very unclear for a new dev looking to extend and implement new counting functionality.

> An alternative is to have some other contract GovernorWithParams where Governor is kept as the backwards compatible version. What I don't like is this adds quite some complexity and due to Solidity override syntax leaking implementation details it is probably a breaking change in another way.

Alright, I went ahead and tried this just to get a sense of what it would be like, and pushed it up to a [new branch](https://github.com/ScopeLift/openzeppelin-contracts/tree/flex-vote-extend). Here's [the diff](https://github.com/ScopeLift/openzeppelin-contracts/commit/b82a777d03ec9cb6387b755cf0ebb8ca77eb3796#). Note that I only took it as far as I needed to get it to compile. It's not cleaned up or documented, and there are failing tests.

Here are a couple of things I notice about this approach:

* I had to duplicate the logic from from the Governor's `_castVote` method in the `GovernorWithParams` new `castVoteWithParams` method, as there's no way to curry to the existing implementation without losing the parameters
* Because of this, I also had to make the proposal state `internal` rather than `private` in order to access it in `GovernorWithParams`

What do you think of this? It feels a bit clunky, but it *could* work.

> Curious if you can think of any other way to make this backwards compatible?

Hmm let me play around a bit more and see if I can come up with something

> If we can't find a good way to make this work we can consider making the breaking change but we strongly prefer not to.

I'm inclined to argue it might be worth making a breaking change here to avoid adding too much technical debt or cruft. In particular, while this is a breaking change for projects that *extend* the contracts, it's not a breaking change from an external interface standpoint. To me that feels less serious, and a reasonable tradeoff to make here. But I completely understand the importance of backwards compatibility as a principle. Not an easy call!

frangio replied:
Thanks for trying this! I agree it's clunky and totally agree it's adding technical debt/cruft which makes no sense for such new contracts.

We've discussed this a bit more and would be happy to make this breaking change since it's super small and contained, unlikely to affect anyone at all at this point, and because it has the potential to enable many more vote counting modules. We just need a short "How to upgrade from 4.5" section in the changelog.

apbendi replied:
Sounds good @frangio. That sounds like the right decision to me! Thanks for consideration. I'll add that to the changelog and also continue with my implementation of #2929 to see what using this extension feels like in practice.

frangio replied:
Thank you! Feel free to share an early draft, I just want to make sure that this mechanism is actually good enough to implement that module.

frangio commented:
Sorry @apbendi I was actually waiting to hear back about your experience implementing the fractional votes before moving forward with this. Just want to make sure we're on the right track with this PR.

apbendi commented:
> Sorry @apbendi I was actually waiting to hear back about your experience implementing the fractional votes before moving forward with this. Just want to make sure we're on the right track with this PR.

Ahh ok, good to know. We have a bit of a chicken/egg problem because I didn't want to go too far with my implementation before knowing the approach would be accepted 😅 

I'll try to break us out of this infinite loop by pushing mine forward a bit soon so I have more concrete feedback to give!

Amxx commented:
Hello @apbendi, I'm joining the conversation to discuss possible additions to this PR.

When a vote is cast there are basically 3 things happening:
1. getting the weight associated with the vote
2. making sure that weight is not re-used to perform another vote latter
3. aggregate that weight to the voting result.

Operation 1 is done in the "Votes" modules, usually, be checking the snapshotted balance of a Comp-like token
Operation 2 and 3 are performed by the "Counting" module.

While your PR proposes to pass an additional "params" to the Counting module, I personally believe it should also be passed to the Votes module. I created [this commit](https://github.com/Amxx/openzeppelin-contracts/commit/f6dfb492fda1ecc7c419a23355111891fd304856) showing how I would do that, and how I think it could be used for voting with non-snapshotted NFT contracts.

What do you think?

Amxx commented:
I also think we may want a `castVoteWithReasonAndParamsBySig`. Not having it would kill the "vote with signature" (metatx) option when params are required.

apbendi commented:
Hi @Amxx 👋 

Great idea! I was going to ask for an example of when this data my needed by Votes module, but your ERC721 implementation provided a perfect example :)

I see no issue with this approach. How would you like to go about handling the merge dependencies here? Shall I rebase this branch to master and apply just the `getVotes` related changes onto this branch?

> I also think we may want a castVoteWithReasonAndParamsBySig. Not having it would kill the "vote with signature" (metatx) option when params are required.

I have no issue with this, and believe I may have included originally but removed it based on [this comment](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043#issuecomment-999207558). Only one follow up question: are there any concerns about contract size constraints?

Amxx commented:
You can fast forward your branch to my commit, or just apply it. Whatever works best for you.

For the size constraints, it could become an issue ... but I'd argue that as long as the mock can be deployed, that is good. In the worst case, we should probably cut revert strings before we remove features.

Amxx commented:
> Cool, so do we want to include the ERC721 voting implementation in this PR?

@frangio what do you think?

frangio commented:
No I wouldn't include that yet. We should have a small mock for testing the params though.

apbendi commented:
@frangio I modified @Amxx's commit to remove the 721 implementation and also rebased this whole branch to master. I'll work on getting a mock and some proper tests set up. Also, to confirm, I should (re)add the `castVoteWithReasonAndParamsBySig` method?

Amxx commented:
Thank you a lot @apbendi 

IMO yes, we should re add `castVoteWithReasonAndParamsBySig`

Amxx commented:
When we get to it, we'll have to add a breaking change notice (override change)

apbendi commented:
Hey @frangio and @Amxx, I've created Mock and used it to test usage of the Governance extension parameters. I've also implemented `castVoteWithReasonAndParamsBySig` and included in the tests. Could you take a quick look and tell me if everything looks good to you?

A couple of additional questions:

* Should we consider adding `params` to the `VoteCast` event emitted by the Governor?
* What do I need to do to update the changelog/breaking change notice? Just an entry in the "Unreleased" section at the top of the `CHANGELOG.md`?

apbendi commented:
Hey @frangio and @Amxx, friendly bump here :)

frangio commented:
Ohh the event. That's annoying. Adding params to `VoteCast` is a breaking change. If we want to log params in an event (and we probably should), we need to emit a separate event, but how should a consumer of the events link it to the corresponding `VoteCast`? And do we always emit it or for example only if `params.length > 0`?

For the changelog just a description in Unreleased, yes.

Mock and tests look good.

frangio commented:
Another idea to avoid having to merge data from two events: emit the current `VoteCast` only if `params.length == 0`, and emit a new event with params if `params.length > 0`.

I think this is better. It drops compatibility with GovernorBravo events but that doesn't make any sense when you introduce params because the vote can't be interpreted unless you know them.

apbendi commented:
Hey @frangio and @Amxx, thanks for your patience here! Had to take a short break from this project but picking it back up in earnest now.

I believe I've made the changes you recommended [here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043#issuecomment-1021196029). I've also taken a pass at adding everything to the Changelog. At this point I believe the PR is ~complete, so I'm taking it out of Draft for your full review.

One note I wanted to call out is just to revisit the naming. "Params" still feels kind of clunky, and it collides with params, i.e. normal function parameters. I don't have any better suggestions, but just wanted to mention it one last time to see if you had any other ideas. One option would be to go back to `data`, but as previously discussed I don't think that's great either! Happy to hear thoughts on this.

Otherwise, I'm looking forward to your full feedback. Thanks a ton for working with us to push this forward :)

frangio commented:
Alternatives to params could be:
- spec / specs (as in specification)
- details

apbendi commented:
Hmm, good ideas but they feel about equivalent to params. Some other ideas for you @frangio:

* info
* memo
* note

apbendi commented:
> * info
> * memo
> * note

@frangio @Amxx let me know if you prefer any of these options over `params` and I'll make the change. Otherwise I'm fine with sticking with `params` as is.

Also curious to hear what you think of the rest of the PR and if any further changes are needed! Thanks :)

frangio commented:
Hi @apbendi, I will be reviewing this in more depth later today. But based on a quick look and our previous discussion I don't expect that much more work will be necessary.

For now "params" is the term that has stuck the most with me... The concept of "parameterized voting" just makes a lot of sense to me. But I'll give this some more thought.

In your opinion would you say these changes are being tested appropriately? We know that the Governor tests are not easy to work with and have some improvements planned, but we have to work with this style of test for now.

We plan to include this in the next release. Thank you for the patience!!

apbendi commented on CHANGELOG.md:
```diff
@@ -4,6 +4,17 @@
+ * `Governor`: Add default `getVotes` implementation which calls virtual internal `_getVotes` method. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
+ * `Governor`: Add `EXTENDED_BALLOT_TYPEHASH` used for signature validation in `castVoteWithReasonAndParamsBySig`. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
+
+### Breaking changes
```
Done in 3e27efdaf1dd78d25be44ff5dbd198bc3d28916c

apbendi commented on contracts/governance/Governor.sol:
```diff
@@ -299,12 +329,19 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
+    /**
+     * @dev See {IGovernor-getVotes}.
+     */
+    function getVotes(address account, uint256 blockNumber) public view virtual override returns (uint256) {
```
done in 5a905ad5519efdc56cbe3b2946ad3484a9401456

apbendi replied:
@frangio I'm realizing something as I look at this now: shouldn't we also be adding a public `getVotes(address account, uint256 blockNumber, bytes memory params)` version of this function as well? As it stands, there would be no way to externally get the vote count given a set of `params` to be included in the vote, unless I'm missing something.

frangio replied:
Hm, yes this sounds like it would be important to build a UI.

Amxx replied:
Do we want to add a `getVotesWithParams` public view ?

apbendi replied:
Done in 455efe84dc5a8f2f1cd4481961722f880775f1d5

apbendi commented on contracts/governance/compatibility/GovernorCompatibilityBravo.sol:
```diff
@@ -265,8 +265,11 @@ abstract contract GovernorCompatibilityBravo is IGovernorTimelock, IGovernorComp
+        uint256 weight,
+        bytes memory params
     ) internal virtual override {
+        (params); // silence compiler warnings
```
Done in 2959ba620a05d51c336b37b4711c27541bb2a5b8

apbendi commented on contracts/governance/Governor.sol:
```diff
@@ -299,12 +329,19 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
     function castVote(uint256 proposalId, uint8 support) public virtual override returns (uint256) {
         address voter = _msgSender();
-        return _castVote(proposalId, voter, support, "");
+        return _castVote(proposalId, voter, support, "", _defaultParams());
```
Done in 6f65ba7eb6fcf90437b9857c48044b471406d1ba

apbendi commented on contracts/governance/Governor.sol:
```diff
@@ -175,6 +185,15 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
     function _voteSucceeded(uint256 proposalId) internal view virtual returns (bool);
 
+    /**
+     * @dev Check the voting weight of an account.
```
Done in 5a905ad5519efdc56cbe3b2946ad3484a9401456

apbendi commented on CHANGELOG.md:
```diff
@@ -4,6 +4,17 @@
+ * `IGovernor`: Add a new event `VoteCastWithParams` that is emitted when a vote is cast that includes non-empty params data. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
+ * `Governor`: Add internal `_defaultParams` method and default implementation. The return value is passed as params data when `castVote` methods without params are called. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
+ * `Governor`: Add default `getVotes` implementation which calls virtual internal `_getVotes` method. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
+ * `Governor`: Add `EXTENDED_BALLOT_TYPEHASH` used for signature validation in `castVoteWithReasonAndParamsBySig`. ([#3043](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043))
```
Done in 3e27efdaf1dd78d25be44ff5dbd198bc3d28916c

apbendi commented:
Thanks @frangio! I've rebased to resolve the changelog merge conflict and addressed all your feedback. I also have one [new question](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#r807355433) for you that occurred to me while making the changes.

> For now "params" is the term that has stuck the most with me... The concept of "parameterized voting" just makes a lot of sense to me. But I'll give this some more thought.

Sounds good. I'm fine with `params` if you're comfortable with it. Felt worthwhile to do one last sanity check before committing to it.

> In your opinion would you say these changes are being tested appropriately? We know that the Governor tests are not easy to work with and have some improvements planned, but we have to work with this style of test for now.

They were a little tricky to grok but I think I've got my head around them and I think we're doing an adequate job exercising them:

* We're using a [mock](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-781636e03e7bc2a753d8f1a0115f476c10d219e41408033a0e18b360daf8f50eR8) that emits an arbitrary even with `params` iff they're included.
* We make sure the [nominal](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-1f4f589c1d290b0bb99231af518d3eeb7c87a46e91a1ec8d0358379158462461R45) use case is still working.
* We make sure that voting with params [is working](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-1f4f589c1d290b0bb99231af518d3eeb7c87a46e91a1ec8d0358379158462461R111)
* ....along with voting with params [via signature](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-1f4f589c1d290b0bb99231af518d3eeb7c87a46e91a1ec8d0358379158462461R149).
* The mock [also implements](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-781636e03e7bc2a753d8f1a0115f476c10d219e41408033a0e18b360daf8f50eR25) `_getVotes` to arbitrarily reduce the voter's weight, and we exercise and [check](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3043/files#diff-1f4f589c1d290b0bb99231af518d3eeb7c87a46e91a1ec8d0358379158462461R144) for this reduction after the vote has occurred.

frangio commented on contracts/mocks/GovernorWithParamsMock.sol:
```diff
@@ -0,0 +1,61 @@
+            (reduction, ) = abi.decode(params, (uint256, string));
+        }
+        // reverts on overflow
+        return GovernorVotes._getVotes(account, blockNumber, params) - reduction;
```
Keep in mind that not using `super` is generally a mistake.

```suggestion
        return super._getVotes(account, blockNumber, params) - reduction;
```

apbendi replied:
Fixed in 8722c5cade8f3de36bebd14aa1ffdc50cf32572d

frangio commented on contracts/mocks/GovernorWithParamsMock.sol:
```diff
@@ -0,0 +1,61 @@
+            (uint256 _uintParam, string memory _strParam) = abi.decode(params, (uint256, string));
+            emit CountParams(_uintParam, _strParam);
+        }
+        return GovernorCountingSimple._countVote(proposalId, account, support, weight, params);
```
```suggestion
        return super._countVote(proposalId, account, support, weight, params);
```

apbendi replied:
Fixed in 8722c5cade8f3de36bebd14aa1ffdc50cf32572d

frangio commented on test/governance/extensions/GovernorWithParams.test.js:
```diff
@@ -0,0 +1,233 @@
+      const tx = await this.mock.castVoteWithReasonAndParams(this.id, Enums.VoteType.For, '', params, { from: voter2 });
+
+      expectEvent(tx, 'CountParams', { uintParam, strParam });
+      expectEvent(tx, 'VoteCastWithParams', { voter: voter2, weight: reducedWeight, params });
```
Can we add an assertion using the `proposalVotes` getter to see more directly that the votes were counted considering the params?

apbendi replied:
Good idea! Done here with b4cd7c4caaf5f87a445125cb6be117857559d5ce

frangio commented:
Can you add a mention of `params` in the [docs for `COUNTING_MODE`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/6fec54f17f0bcdd8d2813ff280c1f385f0796e19/contracts/governance/IGovernor.sol#L71)? If a counting module makes use of encoded params, it should include this in the `COUNTING_MODE` string under a new `params` key with a unique name that describes the behavior.

---

This is looking good to me. I'd like @Amxx to give this a review as well, but I don't really have further comments. I will want to get some external feedback before releasing this but I'm ready to merge it after these last few comments are resolved.

apbendi commented:
hey @frangio and @Amxx, thanks again for your patience. I'm back from ETHDenver, recovered from Covid, and finally had a chance to address your latest comments. I believe everything has been resolved. Please let me know if anything else is needed before this can be merged. Can't say it enough: really appreciate your feedback through this process!

apbendi commented:
hey @frangio friendly bump on this PR :)

frangio reviewed (approved):
LGTM. Thank you for the patience and working with us!
